### PR TITLE
feat(copilot): add GitHub Copilot usage provider

### PR DESCRIPTION
**Pull request by robinebers**
## TL;DR

Adds a native **GitHub Copilot** provider (Premium / Chat / Completions quota meters) that reads a GitHub token already on the machine — no login flow and, by design, no browser cookies. Ports the legacy Tauri plugin and layers in the more robust response parsing from CodexBar.

## What was happening

- The legacy (Tauri) edition shipped a Copilot plugin, but the Swift edition had no Copilot provider at all.
- The legacy plugin's parsing was thin: it didn't handle zero-entitlement "placeholder" buckets (token-based-billing seats), `unlimited` buckets, or the `yyyy-MM-dd` reset-date shape — so it could render misleading meters for some accounts.

## What this changes

- New provider under `Sources/OpenUsage/Providers/Copilot/` (auth store, usage client, mapper, runtime), conforming to `ProviderRuntime`.
- **Auth (read existing creds, no cookies):** first non-empty of — Copilot editor `~/.config/github-copilot/apps.json` (older `hosts.json`) → `gh` `~/.config/gh/hosts.yml` `oauth_token` → `gh` macOS keychain (`gh:github.com`, go-keyring-base64 decode).
- **API:** `GET https://api.github.com/copilot_internal/user` with the standard Copilot client headers (API version `2025-04-01`, `Authorization: token`).
- **Mapping (robustness ported from CodexBar, 14K★):** suppresses zero-entitlement placeholder buckets instead of rendering a misleading "100% used"; treats `unlimited` buckets as empty meters with no reset; parses ISO **or** bare `yyyy-MM-dd` reset dates; supports both paid (`quota_snapshots`) and free (`limited_user_quotas`/`monthly_quotas`) tiers; surfaces an honest "token-based billing — usage not available" message only when nothing is metrically usable.
- **Metrics & defaults:** Premium / Chat / Completions percent meters (ids `copilot.premium|chat|completions`). Premium is primary **and** pinned; Chat + Completions are secondary (behind the caret); provider enabled by default.
- Registered in `AppContainer` (alphabetical: after Antigravity, before Devin), error categories wired in `ErrorCategory.swift`, icon ported from legacy (`Resources/ProviderIcons/copilot.svg`), and docs added at `docs/providers/copilot.md`.

## Heads-up

- **No cookie reader, by design.** GitHub only exposes Copilot dollar/spend data via its logged-in web billing page (what CodexBar reads via an opt-in cookie feature). This is intentionally out of scope — OpenUsage will not read browser cookies — so there are no spend tiles for Copilot.
- **Token-based billing is not the same as "no data."** Verified against a real account where `token_based_billing: true` coexists with real Chat/Completions quotas and a zero-entitlement Premium bucket. The mapper only shows the "not available" message when *no* bucket is usable; otherwise it shows the real meters and drops the empty Premium bucket.
- For accounts with zero premium entitlement, the pinned Premium tile reads "No data" (honest) rather than a fabricated number — easy to revisit if a different default is preferred.

## Tests

- 16 new tests in `Tests/OpenUsageTests/CopilotProviderTests.swift` (auth-source precedence + go-keyring decode, paid/free/unlimited/placeholder mapping, reset-date parsing, and provider error paths). Full suite: **507 pass, 0 failures**.
- Verified end-to-end in the running app via the local usage API (`/v1/usage/copilot`): plan **Individual**, **Chat 9% used**, **Completions ~1% used** (reset 2026-07-01), Premium correctly omitted.

## Screenshots

image

The provider renders the same meter-tile components as existing providers (e.g. Devin/Grok). Live in-app output from the running dev build:

```
plan: Individual
  Chat:        9.0% used   resets 2026-07-01
  Completions: 0.6% used   resets 2026-07-01
  Premium:     (No data — zero premium entitlement on this account)
```

---

> [!NOTE]
> **Medium Risk**
> Reads local OAuth tokens and Keychain and calls GitHub’s internal Copilot API; behavior is isolated like other providers and heavily tested, but credential handling and API compatibility (e.g. hardcoded client version headers) can affect real users over time.
> 
> **Overview**
> Adds a **GitHub Copilot** provider that shows **Premium**, **Chat**, and **Completions** as percent-used meters with plan name and reset countdown when GitHub returns one.
> 
> **Auth** reuses tokens already on disk—Copilot editor `apps.json`/`hosts.json`, then `gh` `hosts.yml`, then macOS Keychain (`gh:github.com`)—with **github.com** scoping so Enterprise tokens are not sent to `api.github.com`. **Usage** calls `GET /copilot_internal/user` with Copilot-style headers and `Authorization: token …`.
> 
> **Mapping** converts GitHub’s percent-remaining into percent-used for paid `quota_snapshots` and free `limited_user_quotas`/`monthly_quotas`; it drops zero-entitlement placeholders, handles `unlimited` buckets, and for **token-based billing** returns plan-only empty meters instead of errors.
> 
> Wired into **`AppContainer`**, **`DefaultLayout`** (Premium pinned; Chat/Completions behind expand), **`ErrorCategory`**, provider icon, **`docs/providers/copilot.md`**, and broad unit/provider tests.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 376b9ad6697ca267930a905b694a8949590d3dcb. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

<h3>Confidence Score: 5/5</h3>

Safe to merge — new provider is fully isolated, degrades gracefully on auth failures, and all edge-case parsing paths are unit-tested.

The credential sourcing logic is carefully scoped (multi-host GHE issue was addressed), the mapper handles every documented edge case with explicit tests, and error paths all surface friendly messages rather than crashing or silently returning stale data. The PR follows all established provider conventions.

No files require special attention.

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `Sources/OpenUsage/Providers/Copilot/CopilotUsageClient.swift`, line 282-292 ([link](https://github.com/robinebers/openusage/blob/c2c41816468ea896f4dd32c5795ff5276f901e11/Sources/OpenUsage/Providers/Copilot/CopilotUsageClient.swift#L282-L292)) 

   <a href="#">P2</a> **Hardcoded editor version strings will drift over time**

   `Editor-Version`, `Editor-Plugin-Version`, and `User-Agent` are pinned to specific VS Code and Copilot Chat builds (`vscode/1.96.2`, `copilot-chat/0.26.7`). GitHub has historically rejected Copilot API calls from clients reporting very old plugin versions. There's no mechanism here to update these without a code release. A comment noting where the authoritative values can be found (e.g., the VS Code Copilot extension's `package.json`) and how frequently they should be bumped would help future maintainers keep up.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: Sources/OpenUsage/Providers/Copilot/CopilotUsageClient.swift
   Line: 282-292

   Comment:
   **Hardcoded editor version strings will drift over time**

   `Editor-Version`, `Editor-Plugin-Version`, and `User-Agent` are pinned to specific VS Code and Copilot Chat builds (`vscode/1.96.2`, `copilot-chat/0.26.7`). GitHub has historically rejected Copilot API calls from clients reporting very old plugin versions. There's no mechanism here to update these without a code release. A comment noting where the authoritative values can be found (e.g., the VS Code Copilot extension's `package.json`) and how frequently they should be bumped would help future maintainers keep up.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!

   <a href="https://app.greptile.com/api/ide/cursor?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20Sources%2FOpenUsage%2FProviders%2FCopilot%2FCopilotUsageClient.swift%0ALine%3A%20282-292%0A%0AComment%3A%0A**Hardcoded%20editor%20version%20strings%20will%20drift%20over%20time**%0A%0A%60Editor-Version%60%2C%20%60Editor-Plugin-Version%60%2C%20and%20%60User-Agent%60%20are%20pinned%20to%20specific%20VS%20Code%20and%20Copilot%20Chat%20builds%20%28%60vscode%2F1.96.2%60%2C%20%60copilot-chat%2F0.26.7%60%29.%20GitHub%20has%20historically%20rejected%20Copilot%20API%20calls%20from%20clients%20reporting%20very%20old%20plugin%20versions.%20There's%20no%20mechanism%20here%20to%20update%20these%20without%20a%20code%20release.%20A%20comment%20noting%20where%20the%20authoritative%20values%20can%20be%20found%20%28e.g.%2C%20the%20VS%20Code%20Copilot%20extension's%20%60package.json%60%29%20and%20how%20frequently%20they%20should%20be%20bumped%20would%20help%20future%20maintainers%20keep%20up.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&pr=764&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursorDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursor.svg?v=3">Fix in Cursor</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20Sources%2FOpenUsage%2FProviders%2FCopilot%2FCopilotUsageClient.swift%0ALine%3A%20282-292%0A%0AComment%3A%0A**Hardcoded%20editor%20version%20strings%20will%20drift%20over%20time**%0A%0A%60Editor-Version%60%2C%20%60Editor-Plugin-Version%60%2C%20and%20%60User-Agent%60%20are%20pinned%20to%20specific%20VS%20Code%20and%20Copilot%20Chat%20builds%20%28%60vscode%2F1.96.2%60%2C%20%60copilot-chat%2F0.26.7%60%29.%20GitHub%20has%20historically%20rejected%20Copilot%20API%20calls%20from%20clients%20reporting%20very%20old%20plugin%20versions.%20There's%20no%20mechanism%20here%20to%20update%20these%20without%20a%20code%20release.%20A%20comment%20noting%20where%20the%20authoritative%20values%20can%20be%20found%20%28e.g.%2C%20the%20VS%20Code%20Copilot%20extension's%20%60package.json%60%29%20and%20how%20frequently%20they%20should%20be%20bumped%20would%20help%20future%20maintainers%20keep%20up.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=robinebers%2Fopenusage&pr=764&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=3">Fix in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22robinebers%2Fopenusage%22%20on%20the%20existing%20branch%20%22claude%2Fjovial-germain-734c0b%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22claude%2Fjovial-germain-734c0b%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20Sources%2FOpenUsage%2FProviders%2FCopilot%2FCopilotUsageClient.swift%0ALine%3A%20282-292%0A%0AComment%3A%0A**Hardcoded%20editor%20version%20strings%20will%20drift%20over%20time**%0A%0A%60Editor-Version%60%2C%20%60Editor-Plugin-Version%60%2C%20and%20%60User-Agent%60%20are%20pinned%20to%20specific%20VS%20Code%20and%20Copilot%20Chat%20builds%20%28%60vscode%2F1.96.2%60%2C%20%60copilot-chat%2F0.26.7%60%29.%20GitHub%20has%20historically%20rejected%20Copilot%20API%20calls%20from%20clients%20reporting%20very%20old%20plugin%20versions.%20There's%20no%20mechanism%20here%20to%20update%20these%20without%20a%20code%20release.%20A%20comment%20noting%20where%20the%20authoritative%20values%20can%20be%20found%20%28e.g.%2C%20the%20VS%20Code%20Copilot%20extension's%20%60package.json%60%29%20and%20how%20frequently%20they%20should%20be%20bumped%20would%20help%20future%20maintainers%20keep%20up.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=robinebers%2Fopenusage&pr=764&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a>
</details>

<sub>Reviews (3): Last reviewed commit: ["fix(copilot): only use a github.com toke..."](https://github.com/robinebers/openusage/commit/376b9ad6697ca267930a905b694a8949590d3dcb) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=40251826)</sub>

### CODE REVIEW

File: Sources/OpenUsage/Providers/Copilot/CopilotUsageMapper.swift [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            appendIfPresent(&lines, limitedLine(label: "Completions", remaining: limited?["completions"], total: monthly?["completions"], resetsAt: resetsAt))
+        }
+
+        guard !lines.isEmpty else {
+            // Distinguish "Business seat with no per-seat quota" from a generic empty payload so the
+            // error badge can say something true instead of looking like a malfunction.
+            if readBool(body["token_based_billing"]) == true {
+                throw CopilotUsageError.tokenBasedBilling
+            }
+            throw CopilotUsageError.quotaUnavailable
```
[the remark this answers is not in the document]
Comment (robinebers) [reply Sources/OpenUsage/Providers/Copilot/CopilotUsageMapper.swift:53]: Fixed in d8ad6dd — token-based-billing now returns a success snapshot carrying the plan with empty meters (tiles read "No data") instead of a plan-less, uncached error, so the plan badge shows. Genuinely empty/garbled payloads (no token-based marker) still fail loudly with `quotaUnavailable`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: Sources/OpenUsage/Providers/Copilot/CopilotAuthStore.swift [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            guard let data = Data(base64Encoded: encoded),
+                  let decoded = String(data: data, encoding: .utf8)
+            else {
+                return nil
+            }
+            text = decoded.trimmingCharacters(in: .whitespacesAndNewlines)
+        }
+        return text.nilIfEmpty
+    }
+}
```
[the remark this answers is not in the document]
Comment (robinebers) [reply Sources/OpenUsage/Providers/Copilot/CopilotAuthStore.swift:186]: Fixed in d8ad6dd — `yamlValue` now scopes the scan to the `github.com:` host block, so an earlier GitHub Enterprise block can't leak its `oauth_token` to api.github.com. Added multi-host tests.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: Sources/OpenUsage/Providers/Copilot/CopilotAuthStore.swift [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            }
+            return token
+        }
+
+        for (key, value) in object where key == "github.com" || key.hasPrefix("github.com:") {
+            if let token = token(in: value) { return token }
+        }
+        for (_, value) in object {
+            if let token = token(in: value) { return token }
+        }
```
[the remark this answers is not in the document]
Comment (robinebers) [reply Sources/OpenUsage/Providers/Copilot/CopilotAuthStore.swift]: Fixed in 376b9ad — `oauthToken(fromEditorJSON:)` now only returns a `github.com` token (exact `github.com` or `github.com:<appId>` keys); the any-host fallback is gone, so a non–github.com editor entry returns nil and the chain falls through to gh config / keychain. Added tests for the Enterprise-only and mixed-host cases.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
